File: sections/70_bounded_min_cost_flows/problems/a_bounded_circulation/solution.cpp
#include <bits/stdc++.h>
using namespace std;
using int64 = long long;

struct Dinic {
  struct Edge {
    int to, rev;
    int64 cap;
  };
  vector<vector<Edge>> g;
  vector<int> level, it;
  Dinic(int n) : g(n), level(n), it(n) {}
  void add_edge(int u, int v, int64 cap) {
    Edge a{v, (int)g[v].size(), cap};
    Edge b{u, (int)g[u].size(), 0};
    g[u].push_back(a);
    g[v].push_back(b);
  }
  bool bfs(int s, int t) {
    fill(level.begin(), level.end(), -1);
    queue<int> q;
    level[s] = 0;
    q.push(s);
    while (!q.empty()) {
      int u = q.front();
      q.pop();
      for (auto &e : g[u])
        if (e.cap && level[e.to] < 0)
          level[e.to] = level[u] + 1, q.push(e.to);
    }
    return level[t] >= 0;
  }
  int64 dfs(int u, int t, int64 pushed) {
    if (u == t)
      return pushed;
    for (int &i = it[u]; i < (int)g[u].size(); ++i) {
      Edge &e = g[u][i];
      if (!e.cap || level[e.to] != level[u] + 1)
        continue;
      int64 take = dfs(e.to, t, min(pushed, e.cap));
      if (take) {
        e.cap -= take;
        g[e.to][e.rev].cap += take;
        return take;
      }
    }
    return 0;
  }
  int64 flow(int s, int t) {
    int64 ans = 0, pushed;
    while (bfs(s, t)) {
      fill(it.begin(), it.end(), 0);
      while ((pushed = dfs(s, t, (1LL << 62))))
        ans += pushed;
    }
    return ans;
  }
};

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  int n, m;
  if (!(cin >> n >> m))
    return 0;
  int ss = n, tt = n + 1;
  Dinic dinic(n + 2);
  vector<int64> balance(n);
  for (int i = 0, u, v; i < m; ++i) {
    int64 low, high;
    cin >> u >> v >> low >> high;
    --u;
    --v;
    dinic.add_edge(u, v, high - low);
    balance[u] -= low;
    balance[v] += low;
  }
  int64 need = 0;
  for (int v = 0; v < n; ++v) {
    if (balance[v] > 0)
      dinic.add_edge(ss, v, balance[v]), need += balance[v];
    if (balance[v] < 0)
      dinic.add_edge(v, tt, -balance[v]);
  }
  cout << (dinic.flow(ss, tt) == need ? "YES\n" : "NO\n");
}

File: sections/70_bounded_min_cost_flows/problems/a_bounded_circulation/solution.py
import collections
import sys


class Dinic:
    def __init__(self, n):
        self.g = [[] for _ in range(n)]

    def add_edge(self, u, v, cap):
        self.g[u].append([v, cap, len(self.g[v])])
        self.g[v].append([u, 0, len(self.g[u]) - 1])

    def flow(self, s, t):
        answer = 0
        while True:
            level = [-1] * len(self.g)
            level[s] = 0
            q = collections.deque([s])
            while q:
                u = q.popleft()
                for v, cap, _ in self.g[u]:
                    if cap and level[v] < 0:
                        level[v] = level[u] + 1
                        q.append(v)
            if level[t] < 0:
                return answer
            it = [0] * len(self.g)

            def dfs(u, pushed):
                if u == t:
                    return pushed
                while it[u] < len(self.g[u]):
                    e = self.g[u][it[u]]
                    v, cap, rev = e
                    if cap and level[v] == level[u] + 1:
                        take = dfs(v, min(pushed, cap))
                        if take:
                            e[1] -= take
                            self.g[v][rev][1] += take
                            return take
                    it[u] += 1
                return 0

            while pushed := dfs(s, 10**30):
                answer += pushed


def main():
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    n, m = data[:2]
    dinic = Dinic(n + 2)
    balance = [0] * n
    at = 2
    for _ in range(m):
        u, v, low, high = data[at : at + 4]
        at += 4
        u -= 1
        v -= 1
        dinic.add_edge(u, v, high - low)
        balance[u] -= low
        balance[v] += low
    need = 0
    for v, value in enumerate(balance):
        if value > 0:
            dinic.add_edge(n, v, value)
            need += value
        elif value < 0:
            dinic.add_edge(v, n + 1, -value)
    print("YES" if dinic.flow(n, n + 1) == need else "NO")


if __name__ == "__main__":
    main()
